File: Scripts/Functions/PerformanceCalculationFunctions.R
# Zero rate Classification for dichotomous models -------------------------
calculate_zero_rate_baseline <- function(ground_truth_labels, model_type, target_class = NULL) {
  if (model_type == "OCC" | model_type == "Binary") {
    # For OCC and Binary, always predict the target class
    zero_rate_preds <- factor(
      rep(target_class, length(ground_truth_labels)),
      levels = levels(ground_truth_labels)
    )
    # Calculate metrics
    f1 <- MLmetrics::F1_Score(y_true = ground_truth_labels, y_pred = zero_rate_preds, positive = target_class)
    precision <- MLmetrics::Precision(y_true = ground_truth_labels, y_pred = zero_rate_preds, positive = target_class)
    recall <- MLmetrics::Recall(y_true = ground_truth_labels, y_pred = zero_rate_preds, positive = target_class)
    accuracy <- MLmetrics::Accuracy(y_true = ground_truth_labels, y_pred = zero_rate_preds)
    
  } else {
    print("this doesn't work for multiclass")
  }
  # Return metrics
  c(
    F1_Score = f1,
    Precision = precision,
    Recall = recall,
    Accuracy = accuracy
  )
}


# Unadjusted performance of model -----------------------------------------
multiclass_class_metrics <- function(ground_truth_labels, predictions) { 
  # Ensure predictions and ground_truth_labels are factors with the same levels
  unique_classes <- sort(union(predictions, ground_truth_labels))
  predictions <- factor(predictions, levels = unique_classes)
  ground_truth_labels <- factor(ground_truth_labels, levels = unique_classes)
  
  # Calculate class prevalence
  prevalence <- table(ground_truth_labels) / length(ground_truth_labels)
  prevalence_df <- data.frame(Class = names(prevalence), Prevalence = as.numeric(prevalence))
  
  # Calculate per-class metrics
  class_metrics <- lapply(unique_classes, function(class) {
    binary_true <- factor(ground_truth_labels == class, levels = c(FALSE, TRUE))
    binary_pred <- factor(predictions == class, levels = c(FALSE, TRUE))
    
    c(
      Class = as.character(class),
      F1_Score = MLmetrics::F1_Score(y_true = binary_true, y_pred = binary_pred, positive = TRUE),
      Precision = MLmetrics::Precision(y_true = binary_true, y_pred = binary_pred, positive = TRUE),
      Recall = MLmetrics::Recall(y_true = binary_true, y_pred = binary_pred, positive = TRUE),
      Accuracy = MLmetrics::Accuracy(y_true = binary_true, y_pred = binary_pred)
    )
  })
  
  # Convert metrics to data frame
  class_metrics_df <- do.call(rbind, lapply(class_metrics, function(x) {
    as.data.frame(t(x), stringsAsFactors = FALSE)
  })) %>%
    mutate(across(c(F1_Score, Precision, Recall, Accuracy), as.numeric))
  
  # Merge with prevalence data
  class_metrics_df <- class_metrics_df %>%
    left_join(prevalence_df, by = "Class")
  
  is.nan.data.frame <- function(x) { # sourced from stack excahnge
    do.call(cbind, lapply(x, is.nan))
  }
  class_metrics_df[is.nan.data.frame(class_metrics_df)] <- 0
  
  # Calculate weighted averages
  weighted_metrics <- class_metrics_df %>%
    summarize(across(c(F1_Score, Precision, Recall, Accuracy), ~ sum(.x * prevalence_df$Prevalence, na.rm = TRUE))) %>%
    as.list()
  
  # Calculate macro averages
  macro_metrics <- colMeans(replace(class_metrics_df[, c("F1_Score", "Precision", "Recall", "Accuracy")], 
                                    is.na(class_metrics_df[, c("F1_Score", "Precision", "Recall", "Accuracy")]), 
                                    0)) %>% as.list()
  
  return(list(
    macro_metrics = macro_metrics,
    weighted_metrics = weighted_metrics,
    class_metrics = class_metrics_df
  ))
}


# Format the test data for testing model ----------------------------------
prepare_test_data <- function(test_feature_data, selected_features, behaviour = NULL) {
  # Select features and metadata columns
  selected_data <- test_feature_data[, .SD, .SDcols = c(selected_features, "Activity", "Time", "ID")]
  selected_data <- na.omit(as.data.table(selected_data))
  
  # For dichotomous models, convert to binary classification
  if (!is.null(behaviour)) {
    selected_data$Activity <- ifelse(selected_data$Activity == behaviour, behaviour, "Other")
  }
  
  # Extract labels and metadata
  ground_truth_labels <- selected_data$Activity
  time_values <- selected_data$Time
  ID_values <- selected_data$ID
  
  # Get numeric features only
  numeric_data <- selected_data[, !c("Activity", "Time", "ID"), with = FALSE]
  
  # Remove invalid rows
  invalid_rows <- which(!complete.cases(numeric_data) |
                          !apply(numeric_data, 1, function(row) all(is.finite(row))))
  
  if (length(invalid_rows) > 0) {
    numeric_data <- numeric_data[-invalid_rows, , drop = FALSE]
    ground_truth_labels <- ground_truth_labels[-invalid_rows]
    time_values <- time_values[-invalid_rows]
    ID_values <- ID_values[-invalid_rows]
  }
  
  list(
    numeric_data = numeric_data,
    ground_truth_labels = ground_truth_labels,
    time_values = time_values,
    ID_values = ID_values
  )
}

save_results <- function(results, predictions, ground_truth_labels, time_values, ID_values,
                         dataset_name, model_name, base_path) {
  # Save performance metrics
  fwrite(results, file.path(base_path, "Output", "Testing", 
                            paste0(dataset_name, "_", model_name, "_test_performance.csv")))
  
  # Save predictions
  output <- data.table(
    "Time" = time_values,
    "ID" = ID_values,
    "Ground_truth" = ground_truth_labels,
    "Predictions" = predictions
  )
  
  if (nrow(output) > 0) {
    fwrite(output, file.path(base_path, "Output", "Testing", "Predictions",
                             paste(dataset_name, model_name, "predictions.csv", sep = "_")))
  }
}



# Random baseline for multiclass metrics ----------------------------------
random_baseline_metrics <- function(ground_truth_labels, iterations = 100) {
  # Convert ground truth to factor and get class information
  ground_truth_labels <- factor(ground_truth_labels)
  class_levels <- levels(ground_truth_labels)
  class_props <- prop.table(table(ground_truth_labels))
  n_classes <- length(class_levels)
  n_samples <- length(ground_truth_labels)
  
  # Pre-allocate matrices for storing results
  class_metrics <- list()
  # Run iterations
  for (i in 1:iterations) {
    # Generate random predictions
    random_preds <- factor(
      sample(class_levels, size = n_samples, prob = class_props, replace = TRUE),
      levels = class_levels
    )
    
    random_baseline <- multiclass_class_metrics(ground_truth_labels, random_preds)
    random_baseline_class <- random_baseline$class_metrics
    
    # Store class metrics
    class_metrics[[i]] <- random_baseline_class
    }
    
  # Calculate average for each class and metric
  class_metrics_combined <- do.call(rbind, class_metrics)
  
  averages_df <- class_metrics_combined %>%
    group_by(Class) %>%
    summarise(
      F1_Score = mean(F1_Score, na.rm = TRUE),
      Precision = mean(Precision, na.rm = TRUE),
      Recall = mean(Recall, na.rm = TRUE),
      Accuracy = mean(Accuracy, na.rm = TRUE),
      Prevalence = mean(Prevalence, na.rm = TRUE)
    )
  
  # calculate the weighted average from the individual classes
  weighted_metrics <- averages_df %>%
    summarize(across(c(F1_Score, Precision, Recall, Accuracy), ~ sum(.x * averages_df$Prevalence, na.rm = TRUE))) %>%
    as.list()
  
  # Calculate summary statistics
  list(
    macro_summary = list(
      F1_Score = weighted_metrics$F1_Score,
      Precision = weighted_metrics$Precision,
      Recall = weighted_metrics$Recall,
      Accuracy = weighted_metrics$Accuracy
    ),
    class_summary = averages_df
  )
}



# Unadjusted performance, random, and zero baseline -----------------------
calculate_full_multi_performance <- function(ground_truth_labels, predictions){
  
  # 1. absolute scores (unadjusted)
  macro_multiclass_scores <- multiclass_class_metrics(ground_truth_labels, predictions)
  class_metrics_df <- macro_multiclass_scores$class_metrics
  macro_metrics <- macro_multiclass_scores$macro_metrics
  weighted_metrics <- macro_multiclass_scores$weighted_metrics
  
  # 2. Zero Rate baseline (always predict the majority class)
  majority_class <- names(which.max(table(ground_truth_labels)))
  zero_rate_preds <- factor(
    rep(majority_class, length(ground_truth_labels)),
    levels = levels(as.factor(ground_truth_labels))
  )
  zero_rate_baseline <- multiclass_class_metrics(ground_truth_labels, zero_rate_preds)
  macro_metrics_zero <- zero_rate_baseline$macro_metrics
  class_metrics_zero <- zero_rate_baseline$class_metrics
  
  # 3. Random baseline (randomly select in stratified proportion to true data)
  random_multiclass <- random_baseline_metrics(ground_truth_labels, iterations = 100)
  random_macro_summary <- random_multiclass$macro_summary
  random_class_summary <- random_multiclass$class_summary
  
  # Compile results
  macro_results <- data.frame(
    Dataset = dataset_name,
    Model = behaviour_set,
    Activity = "WeightedMacroAverage",
    F1_Score = weighted_metrics["F1_Score"],
    Precision = weighted_metrics["Precision"],
    Recall = weighted_metrics["Recall"],
    Accuracy = weighted_metrics["Accuracy"],
    Random_F1_Score = random_macro_summary$F1_Score,
    Random_Precision = random_macro_summary$Precision,
    Random_Recall = random_macro_summary$Recall,
    Random_Accuracy = random_macro_summary$Accuracy,
    ZeroR_F1_Score = macro_metrics_zero$F1_Score,
    ZeroR_Precision = macro_metrics_zero$Precision,
    ZeroR_Recall = macro_metrics_zero$Recall,
    ZeroR_Accuracy = macro_metrics_zero$Accuracy
  )
  
  activity_results_list <- list()
  # Loop through each unique activity
  for (activity in unique(ground_truth_labels)) {
    # Create dataframe for current activity
    activity_results_list[[activity]] <- data.frame(
      Dataset = dataset_name,
      Model = behaviour_set,
      Activity = activity,
      
      F1_Score = class_metrics_df$F1_Score[class_metrics_df$Class == activity],
      Precision = class_metrics_df$Precision[class_metrics_df$Class == activity],
      Recall = class_metrics_df$Recall[class_metrics_df$Class == activity],
      Accuracy = class_metrics_df$Accuracy[class_metrics_df$Class == activity],
      
      Random_F1_Score = random_class_summary$F1_Score[random_class_summary$Class == activity],
      Random_Precision = random_class_summary$Precision[random_class_summary$Class == activity],
      Random_Recall = random_class_summary$Recall[random_class_summary$Class == activity],
      Random_Accuracy = random_class_summary$Accuracy[random_class_summary$Class == activity],
      
      ZeroR_F1_Score = class_metrics_zero$F1_Score[class_metrics_zero$Class == activity],
      ZeroR_Precision = class_metrics_zero$Precision[class_metrics_zero$Class == activity],
      ZeroR_Recall = class_metrics_zero$Recall[class_metrics_zero$Class == activity],
      ZeroR_Accuracy = class_metrics_zero$Accuracy[class_metrics_zero$Class == activity]
    )
  }
  
  # Combine all results into a single dataframe
  final_class_results <- do.call(rbind, activity_results_list)
  
  # Combine all results
  test_results <- rbind(test_results, macro_results, final_class_results)
}

File: Scripts/Python/PerformanceCalculationFunctions.py
import numpy as np
import pandas as pd
from sklearn.metrics import f1_score, precision_score, recall_score, accuracy_score
from typing import List, Dict, Union, Tuple
import random


def calculate_zero_rate_baseline(ground_truth_labels: List[str], model_type: str, target_class: str = None) -> Dict[str, float]:
    """Calculate zero rate baseline metrics for dichotomous models."""
    if model_type in ["OCC", "Binary"]:
        # For OCC and Binary, always predict the non-target class
        zero_rate_preds = ["Other"] * len(ground_truth_labels)
        
        # Calculate metrics
        f1 = f1_score(ground_truth_labels, zero_rate_preds, pos_label=target_class, average='binary')
        precision = precision_score(ground_truth_labels, zero_rate_preds, pos_label=target_class, average='binary')
        recall = recall_score(ground_truth_labels, zero_rate_preds, pos_label=target_class, average='binary')
        accuracy = accuracy_score(ground_truth_labels, zero_rate_preds)
        
        return {
            'F1_Score': f1,
            'Precision': precision, 
            'Recall': recall,
            'Accuracy': accuracy
        }
    else:
        print("This doesn't work for multiclass, use other one")
        return {}


def multiclass_class_metrics(ground_truth_labels: List[str], predictions: List[str]) -> Dict:
    """Calculate performance metrics for multiclass classification."""
    # Get unique classes
    unique_classes = sorted(list(set(predictions) | set(ground_truth_labels)))
    
    # Calculate class prevalence
    prevalence = pd.Series(ground_truth_labels).value_counts(normalize=True)
    prevalence_df = pd.DataFrame({'Class': prevalence.index, 'Prevalence': prevalence.values})
    
    # Calculate per-class metrics
    class_metrics = []
    for cls in unique_classes:
        binary_true = [1 if x == cls else 0 for x in ground_truth_labels]
        binary_pred = [1 if x == cls else 0 for x in predictions]
        
        metrics = {
            'Class': cls,
            'F1_Score': f1_score(binary_true, binary_pred, average='binary', zero_division=0),
            'Precision': precision_score(binary_true, binary_pred, average='binary', zero_division=0),
            'Recall': recall_score(binary_true, binary_pred, average='binary', zero_division=0),
            'Accuracy': accuracy_score(binary_true, binary_pred)
        }
        class_metrics.append(metrics)
    
    class_metrics_df = pd.DataFrame(class_metrics)
    class_metrics_df = pd.merge(class_metrics_df, prevalence_df, on='Class', how='left')
    
    # Replace NaN with 0
    class_metrics_df = class_metrics_df.fillna(0)
    
    # Calculate weighted averages
    weighted_metrics = {
        metric: np.sum(class_metrics_df[metric] * class_metrics_df['Prevalence'])
        for metric in ['F1_Score', 'Precision', 'Recall', 'Accuracy']
    }
    
    # Calculate macro averages
    macro_metrics = {
        metric: np.mean(class_metrics_df[metric])
        for metric in ['F1_Score', 'Precision', 'Recall', 'Accuracy']
    }
    
    return {
        'macro_metrics': macro_metrics,
        'weighted_metrics': weighted_metrics,
        'class_metrics': class_metrics_df
    }


def prepare_test_data(test_feature_data: pd.DataFrame, selected_features: List[str], behaviour: str = None) -> Dict:
    """Prepare test data for model evaluation."""
    # Select features and metadata columns
    all_cols = selected_features + ['Activity', 'Time', 'ID']
    selected_data = test_feature_data[all_cols].copy()
    selected_data = selected_data.dropna()
    
    # Extract labels and metadata
    ground_truth_labels = selected_data['Activity'].values
    time_values = selected_data['Time'].values
    ID_values = selected_data['ID'].values
    
    # Get numeric features only
    numeric_data = selected_data.drop(['Activity', 'Time', 'ID'], axis=1)
    
    # Remove invalid rows
    valid_mask = numeric_data.notna().all(axis=1) & np.isfinite(numeric_data).all(axis=1)
    numeric_data = numeric_data[valid_mask]
    ground_truth_labels = ground_truth_labels[valid_mask]
    time_values = time_values[valid_mask]
    ID_values = ID_values[valid_mask]
    
    return {
        'numeric_data': numeric_data,
        'ground_truth_labels': ground_truth_labels,
        'time_values': time_values,
        'ID_values': ID_values
    }


def save_results(results: pd.DataFrame, predictions: List[str], ground_truth_labels: List[str], 
                time_values: List, ID_values: List, dataset_name: str, model_name: str, base_path: str):
    """Save model results and predictions."""
    import os
    
    # Save performance metrics
    results.to_csv(os.path.join(base_path, "Output", "Testing", 
                               f"{dataset_name}_{model_name}_test_performance.csv"))
    
    # Save predictions
    output = pd.DataFrame({
        'Time': time_values,
        'ID': ID_values,
        'Ground_truth': ground_truth_labels,
        'Predictions': predictions
    })
    
    if len(output) > 0:
        output.to_csv(os.path.join(base_path, "Output", "Testing", "Predictions",
                                 f"{dataset_name}_{model_name}_predictions.csv"))


def random_baseline_metrics(ground_truth_labels: List[str], iterations: int = 100, model: str = "Multi") -> Dict:
    """Calculate random baseline metrics."""
    # Get class information
    classes = sorted(list(set(ground_truth_labels)))
    class_props = pd.Series(ground_truth_labels).value_counts(normalize=True)
    n_classes = len(classes)
    n_samples = len(ground_truth_labels)
    
    class_prop_equal = [1/n_classes] * n_classes
    
    class_metrics_prev = []
    class_metrics_equal = []
    
    for _ in range(iterations):
        # Generate predictions with true class proportions
        random_preds_prev = np.random.choice(classes, size=n_samples, p=class_props)
        random_preds_equal = np.random.choice(classes, size=n_samples, p=class_prop_equal)
        
        metrics_prev = multiclass_class_metrics(ground_truth_labels, random_preds_prev)
        metrics_equal = multiclass_class_metrics(ground_truth_labels, random_preds_equal)
        
        class_metrics_prev.append(metrics_prev['class_metrics'])
        class_metrics_equal.append(metrics_equal['class_metrics'])
    
    # Calculate averages
    class_metrics_combined_prev = pd.concat(class_metrics_prev)
    class_metrics_combined_equal = pd.concat(class_metrics_equal)
    
    averages_prev = class_metrics_combined_prev.groupby('Class').mean()
    averages_equal = class_metrics_combined_equal.groupby('Class').mean()
    
    if model in ["OCC", "Binary"]:
        selected_metrics_prev = averages_prev[averages_prev.index != "Other"]
        selected_metrics_equal = averages_equal[averages_equal.index != "Other"]
    else:
        # Calculate weighted averages
        selected_metrics_prev = (averages_prev * averages_prev['Prevalence'].values[:, None]).sum()
        selected_metrics_equal = (averages_equal * averages_equal['Prevalence'].values[:, None]).sum()
    
    return {
        'macro_summary': {
            'F1_Score_prev': selected_metrics_prev['F1_Score'],
            'Precision_prev': selected_metrics_prev['Precision'],
            'Recall_prev': selected_metrics_prev['Recall'],
            'Accuracy_prev': selected_metrics_prev['Accuracy'],
            'F1_Score_equal': selected_metrics_equal['F1_Score'],
            'Precision_equal': selected_metrics_equal['Precision'],
            'Recall_equal': selected_metrics_equal['Recall'],
            'Accuracy_equal': selected_metrics_equal['Accuracy']
        },
        'class_summary_prev': averages_prev.reset_index(),
        'class_summary_equal': averages_equal.reset_index()
    }


def calculate_full_multi_performance(ground_truth_labels: List[str], predictions: List[str], model: str) -> pd.DataFrame:
    """Calculate comprehensive performance metrics including baselines."""
    # 1. Absolute scores
    macro_scores = multiclass_class_metrics(ground_truth_labels, predictions)
    class_metrics_df = macro_scores['class_metrics']
    weighted_metrics = macro_scores['weighted_metrics']
    
    # 2. Zero Rate baseline
    majority_class = pd.Series(ground_truth_labels).mode()[0]
    zero_rate_preds = [majority_class] * len(ground_truth_labels)
    zero_rate_baseline = multiclass_class_metrics(ground_truth_labels, zero_rate_preds)
    
    # 3. Random baseline
    random_multiclass = random_baseline_metrics(ground_truth_labels, iterations=100, model=model)
    
    # Calculate false positives
    confusion = pd.crosstab(pd.Series(predictions), pd.Series(ground_truth_labels))
    false_positives = {cls: confusion.loc[cls].sum() - confusion.loc[cls, cls] 
                      for cls in confusion.index}
    
    # Compile results
    results = []
    
    # Add macro average results
    macro_row = {
        'Dataset': dataset_name,
        'Model': model,
        'Activity': 'WeightedMacroAverage',
        'Prevalence': None,
        'FalsePositives': None
    }
    macro_row.update(weighted_metrics)
    macro_row.update({f'ZeroR_{k}': v for k, v in zero_rate_baseline['macro_metrics'].items()})
    macro_row.update(random_multiclass['macro_summary'])
    results.append(macro_row)
    
    # Add per-class results
    for activity in set(ground_truth_labels):
        class_row = {
            'Dataset': dataset_name,
            'Model': model,
            'Activity': activity,
            'Prevalence': class_metrics_df.loc[class_metrics_df['Class'] == activity, 'Prevalence'].iloc[0],
            'FalsePositives': false_positives.get(activity, 0)
        }
        
        # Add metrics
        metrics = class_metrics_df[class_metrics_df['Class'] == activity].iloc[0]
        class_row.update({
            'F1_Score': metrics['F1_Score'],
            'Precision': metrics['Precision'],
            'Recall': metrics['Recall'],
            'Accuracy': metrics['Accuracy']
        })
        
        # Add baseline metrics
        class_row.update({
            f'Random_{k}': v for k, v in random_multiclass['class_summary_prev']
            .loc[random_multiclass['class_summary_prev']['Class'] == activity].iloc[0].items()
            if k not in ['Class']
        })
        
        class_row.update({
            f'ZeroR_{k}': v for k, v in zero_rate_baseline['class_metrics']
            .loc[zero_rate_baseline['class_metrics']['Class'] == activity].iloc[0].items()
            if k not in ['Class', 'Prevalence']
        })
        
        results.append(class_row)
    
    return pd.DataFrame(results)
